src/com/seafile/seadroid2/provider/SeafileProvider.java

```diff
@@ -288,63 +378,126 @@ public AssetFileDescriptor openDocumentThumbnail(String documentId,
+                @Override
+                public void onCancel() {
+                    Log.d(DEBUG_TAG, "openDocumentThumbnail cancelling download");
+                    future.cancel(true);
+                }
+            });
+        }
 
-                        bmp.compress(Bitmap.CompressFormat.PNG, 100, fileStream);
-                        fileStream.close();
+        try {
+            return future.get();
+        } catch (InterruptedException e) {
+            Log.d(DEBUG_TAG, "openDocumentThumbnail cancelled download");
+            throw new FileNotFoundException();
+        } catch (CancellationException e) {
+            Log.d(DEBUG_TAG, "openDocumentThumbnail cancelled download");
+            throw new FileNotFoundException();
+        } catch (ExecutionException e) {
+            Log.d(DEBUG_TAG, "openDocumentThumbnail error during download", e);
+            throw new FileNotFoundException();
+        }
+    }
 
-                    } catch (IOException e) {
-                        Log.d(getClass().getSimpleName(), "could not transfer thumbnail.");
-                    }
+    @Override
+    public String createDocument (String parentDocumentId, String mimeType, String displayName) throws FileNotFoundException {
+        Log.d(DEBUG_TAG, "createDocument: " + parentDocumentId + "; " + mimeType + "; " + displayName);
+
+        String repoId = DocumentIdParser.getRepoIdFromId(parentDocumentId);
+        if (repoId.isEmpty()) {
+            throw new FileNotFoundException();
+        }
+
+        String parentPath = DocumentIdParser.getPathFromId(parentDocumentId);
+        DataManager dm = createDataManager(parentDocumentId);
+
+        try {
+
+            dm.getReposFromServer(); // refresh cache
+            SeafRepo repo = dm.getCachedRepoByID(repoId);
+
+            List<SeafDirent> list = dm.getDirentsFromServer(repoId, parentPath);
+
+            // first check if target already exist. if yes, abort
+            for (SeafDirent e: list) {
+                if (e.getTitle().equals(displayName)) {
+                    throw new SeafException(0, SeadroidApplication.getAppContext().getString(R.string.saf_file_exist));
                 }
+            }
 
-            }.start();
+            if (repo == null || !repo.hasWritePermission()) {
+                throw new SeafException(0, SeadroidApplication.getAppContext().getString(R.string.saf_write_diretory_exception));
+            } else if (mimeType == null) {
+                // bad mime type given by caller
+                throw new SeafException(0, SeadroidApplication.getAppContext().getString(R.string.saf_bad_mime_type));
+            } else if (mimeType.equals(Document.MIME_TYPE_DIR)) {
+                dm.createNewDir(repoId, parentPath, displayName);
+            } else {
+                dm.createNewFile(repoId, parentPath, displayName);
+            }
+
+            // update parent dirent cache
+            dm.getDirentsFromServer(repoId, parentPath);
```

Comment:
I think we should check the network connections before creating or updating files (e.g. `dm.createNewDir` or `dm.createNewFile`, etc.)